### PR TITLE
Pairing StartupParameters

### PR DESCRIPTION
**Pull request by Stuart-Moore**
Fixes # 

Changes proposed in this pull request:
 - New Function Set-DbaStartupParameter
 - Modified Get-DbaStartupParameter  to work with Set-, added Fred's new messaging stuff and made it work with smo server object or string
 - 

How to test this code: 
- [x] To Set - Set-DbaStartupParameter -SqlServer server1\instance1 -SingleUser -TraceFlags 8032,8048
- [x] To Check - Get-DbaStartupParameter -SqlServer server1\instance1  

Has been tested on minimum requirements:
- [ ]  Powershell 3
- [ ]  Windows 7
- [ ]  SQL Server 2000

Has been tested on maximum requirements:
- [ ]  SQL Server vNext
- [ ]  Windows 10
- [ ]  Azure Database

Tests for tester:
- [x] Working/useful help content, including link to command on dbatools web site
- [ ] All examples work as advertised
- [x] Does not contain template content
- [x] Does not contain excessive/unnecessary amounts of comments
- [ ] Works remotely
- [x] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [x] Works with named instances
- [ ] Works with clustered instances
- [ ] Handles offline/read only databases
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

### CODE REVIEW

**Comment by potatoqualitee** [comment]
WOW! Hell yes! That looks exhausting as hell to code; such a relief. Thank you, @Stuart-Moore. @PowerDBAKlaas - can you approve the code requirements and I can do the testing?

File: functions/Get-DbaStartupParameter.ps1 [hunk 1]
```diff
@@ -59,8 +59,17 @@ Logs in to WMI using the ad\sqladmin credential and gathers simplified informati
 			$servercount++
 			try
 			{
-				$instancename = ($servername.Split('\'))[1]
-				Write-Verbose "Attempting to connect to $servername"
+				#$servername, $instancename = ($sqlserver.Split('\'))
+				if ($null -eq $SqlServer.name)
+				{
+					$servername, $instancename = ($sqlserver.Split('\'))
```
Comment (PowerDBAKlaas) [inline functions/Get-DbaStartupParameter.ps1]: Isn't $sqlserver the collection?
Wouldn't it be $servername that can be split?
Or else rename the variable in the foreach ()
Actually, maybe it's the right time to switch to   $instance in $sqlinstance?

File: functions/Get-DbaStartupParameter.ps1 [hunk 2]
```diff
@@ -59,8 +59,17 @@ Logs in to WMI using the ad\sqladmin credential and gathers simplified informati
 			$servercount++
```
Comment (PowerDBAKlaas) [inline functions/Get-DbaStartupParameter.ps1]: why is $servercount here?

File: functions/Get-DbaStartupParameter.ps1 [hunk 3]
```diff
@@ -164,7 +178,8 @@ Logs in to WMI using the ad\sqladmin credential and gathers simplified informati
 			}
 			catch
 			{
-				Write-Warning "$servername`: $_ "
+			#	Write-Message -Level Warning -Message "$servername`: $_ " -silent $silent
+			write-warning "caught error $_"
```
Comment (PowerDBAKlaas) [inline functions/Get-DbaStartupParameter.ps1]: $_ still holds the servername?

File: functions/Set-DbaStartupParameter.ps1 [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+.PARAMETER NoLoggingToWinEvents
+Don't use Windows Application events log
+
+.PARAMETER StartAsNamedInstance
+Allows you to start a named instance of SQL Server
+
+.PARAMETER DisableMonitoring
+
+.PARAMETER IncreasedExtents
+Increases the number of extents that are allocated for each file in a filegroup. T
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: T ?

File: functions/Set-DbaStartupParameter.ps1 [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+.PARAMETER TraceFlagsOverride
+Overrides the default behavious and replaces any existing trace flags. If not trace flags specified will just remove existing ones
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: typo: behavious

File: functions/Set-DbaStartupParameter.ps1 [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+We then change the startup parameters ahead of some work
+
+After the work has been completed, we can push the original startup parameters back to server1\instance1 and resume normal operation
+
+#>
+	[CmdletBinding(SupportsShouldProcess=$true)]
+	param ([parameter(ValueFromPipeline, Mandatory = $true)]
+		[Alias("ServerInstance", "SqlInstance")]
+		[object[]]$SqlServer,
+		[PSCredential]$Credential,
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: I think we need a different type for credentials.

File: functions/Set-DbaStartupParameter.ps1 [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+    {
+        Write-Message -Level VeryVerbose -Message "StartupObject passed in" -silent:$true
+        $NewStartup = $StartUpConfig
+        $TraceFlagsOverride = $true
+    }
+    else
+    {    
+        Write-Message -Level VeryVerbose -Message "Parameters passed in" -silent:$true
+        $NewStartup = $CurrentStartup.PSObject.copy()
+        foreach ($param in ($PsBoundParameters.keys | ?{$_ -in ($NewStartup.PSObject.Properties.name)}))
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: please replace aliases: ? = Where-object

File: functions/Set-DbaStartupParameter.ps1 [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+        {
+            $NewStartup.TraceFlags =  if($CurrentStartup.TraceFlags -eq 'None'){}else{$CurrentStartup.TraceFlags} 
+        }
+        If ($NewStartup.TraceFlags.Length -ne 0)
+        {
+            $ParameterString += (($NewStartup.TraceFlags.split(',') | Foreach {"-T$_"}) -join ';')+";"
+        }
+    }
+
+    $servername, $instancename = ($sqlserver.Split('\'))
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: this assumes there is only one sqlserver

File: functions/Set-DbaStartupParameter.ps1 [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+
+In this example we take a copy of the existing startup configuration of server1\instance1
+
+We then change the startup parameters ahead of some work
+
+After the work has been completed, we can push the original startup parameters back to server1\instance1 and resume normal operation
+
+#>
+	[CmdletBinding(SupportsShouldProcess=$true)]
+	param ([parameter(ValueFromPipeline, Mandatory = $true)]
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: it says 'accept pipeline input', but I see no PROCESS block

File: functions/Set-DbaStartupParameter.ps1 [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+		[Alias("ServerInstance", "SqlInstance")]
+		[object[]]$SqlServer,
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: this says 'accept multiple sql instances', but I see no foreach loop

File: functions/Set-DbaStartupParameter.ps1 [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+            $ParameterString += "-m$SingleUserDetails;"
+        }
+        else
+        {
+            $ParameterString += "-m;"
+        }
+    }
+    if ($NewStartup.NoLoggingToWinEvents)
+    {
+        $ParameterString += "-n";
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: doesn't this one need a ';"' instead of '";'?

File: functions/Set-DbaStartupParameter.ps1 [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+.EXAMPLE 
+Set-DbaStartupParameter -SqlServer server1\instance1 -SingleUser
+
+Will configure the SQL Instance server1\instance1 to startup up in Single User mode at next startup
+	
+.EXAMPLE
+Set-DbaStartupParameter -SqlServer server1\instance1 -SingleUser -TraceFlags 8032,8048
+This will appened Trace Flags 8032 and 8048 to the startup paramters
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: typo: paramters

File: functions/Set-DbaStartupParameter.ps1 [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+
+.EXAMPLE
+Set-DbaStartupParameter -SqlServer server1\instance1 -SingleUser -TraceFlags 8032,8048 -TraceFlagsOverride
+
+This will set Trace Flags 8032 and 8048 to the startup paramters, removing any existing Trace Flags
```
Comment (PowerDBAKlaas) [inline functions/Set-DbaStartupParameter.ps1]: typo: paramters

File: functions/Set-DbaStartupParameter.ps1 (hunk 4 above)
Comment (Stuart-Moore) [inline functions/Set-DbaStartupParameter.ps1]: typo, fixed in next push

File: functions/Set-DbaStartupParameter.ps1 (hunk 5 above)
Comment (Stuart-Moore) [inline functions/Set-DbaStartupParameter.ps1]: fixed in next push

File: functions/Set-DbaStartupParameter.ps1 (hunk 12 above)
Comment (Stuart-Moore) [inline functions/Set-DbaStartupParameter.ps1]: fixed in next push

File: functions/Set-DbaStartupParameter.ps1 (hunk 13 above)
Comment (Stuart-Moore) [inline functions/Set-DbaStartupParameter.ps1]: Copy and paste with typos, they breed fast.

Fixed in next push

File: functions/Set-DbaStartupParameter.ps1 (hunk 7 above)
Comment (Stuart-Moore) [inline functions/Set-DbaStartupParameter.ps1]: fixed in next push

File: functions/Set-DbaStartupParameter.ps1 (hunk 11 above)
Comment (Stuart-Moore) [inline functions/Set-DbaStartupParameter.ps1]: Yes

File: functions/Get-DbaStartupParameter.ps1 (hunk 2 above)
Comment (Stuart-Moore) [inline functions/Get-DbaStartupParameter.ps1]: I didn't write that bit ;)

File: functions/Get-DbaStartupParameter.ps1 (hunk 3 above)
Comment (Stuart-Moore) [inline functions/Get-DbaStartupParameter.ps1]: That's a removed line I believe, the next line writes the error which will be $_.

**Comment by Stuart-Moore** [comment]
Pushed up a typo fixed version.

Do we want the Get command to be able to be sprayed across a number of instances? I can think of cases where it might be great, and plenty where it'd be a BAD thing?

**Comment by PowerDBAKlaas** [comment]
@Stuart-Moore you mean the Set- ( or Add-)?
It's probably safer to stick to one instance.

So, that would mean removing the [] for $sqlserver,
and still need the process{} block

**Comment by Stuart-Moore** [comment]
Fixed the name in the foreach in Get-
Set- now fixed as a single instance option (and has a PROCESS block()

File: functions/Get-DbaStartupParameter.ps1 (hunk 2 above)
Comment (potatoqualitee) [inline functions/Get-DbaStartupParameter.ps1]: that was some of our old crappy code :(

**Comment by potatoqualitee** [comment]
i fixed up get a lil bit, gonna work on set tomorrow!

**Comment by Stuart-Moore** [comment]
Don't need -TraceFlags $null

If it's not set, and -TraceFlagsOverride specified it just blanks them anyway

**Comment by potatoqualitee** [comment]
Oh, good call @Stuart-Moore - that was dumb of me

**Comment by shaneis** [comment]
Passed all the tests that I could throw at it, but can't do a full round of testing until I'm back at my lap <cough>laptop</cough>.

Nice touch with the -TraceFlagOverride, I was worried about that for a second.
This command definitely deserves a 🥃

**Pull request merged**